Fire a DMX signal for custom events

nicou opened the pull request:
Fire DMX signals for significant LORA and DataHub events and incoming jump warnings.

nicou committed:
Fire a DMX event from some custom events

nicou committed:
Allow usage of copilot

plaa commented on src/models/ship.js:
```diff
@@ -92,14 +93,14 @@ export const Beacon = Bookshelf.Model.extend({
+						shipLogger.warning(`Received a distress signal from area Alpha-5-D2-100: ${velianMessage}`, { showPopup: true });
+						return;
+					}
+					shipLogger.success(`Decryption key '${this.get('id')}' successfully decrypted an unknown signal originating from area ${this.related('grid').get('name')}`, { showPopup: true });
```
Onko tarkoituksella muutettu että `veliamMessage` tapauksessa tulee molemmat logitukset + popupit?  Aiemmin tuli vain toinen.

nicou replied:
Ei tule molemmat, koska ensimmäisen logituksen jälkeen on `return`

plaa commented on src/routes/fleet.js:
```diff
@@ -129,6 +130,7 @@ router.post('/:id/destroy', handleAsyncErrors(async (req, res) => {
 	if (!ship) throw new NotFound('Ship not found');
 	if (ship.get('status') === 'Destroyed') throw new BadRequest(`Ship ${id} is already destroyed`);
 	await ship.destroyShip();
+	dmx.fireEvent(dmx.CHANNELS.FleetShipDestroyed);
```
Voiko aluksen tuhota muualtakin?  Olisiko loogisempaa että tämä ammuttaisiin `ship.destroyShip` metodista?

plaa replied:
Mietin myös että onko aluksen tuhoutuminen aina välittömästi Odysseuksen tiedossa?  Onko tilanteita, jossa halutaan tuhota aluksia ilman että tulee infoa?  Vai onko kannassa olevat alukset olennaisesti Odysseuksen senhetkinen käsitys tilanteessa, eikä "todellisuus"?

nicou replied:
> Voiko aluksen tuhota muualtakin?

Ei tietääkseni

> Olisiko loogisempaa että tämä ammuttaisiin `ship.destroyShip` metodista?

Olisi, muutan tän 😄

> Onko tilanteita, jossa halutaan tuhota aluksia ilman että tulee infoa?

Ei ole kuulemma

plaa commented on src/routes/vote.js:
```diff
@@ -107,6 +108,9 @@ router.put('/:id', handleAsyncErrors(async (req, res) => {
 			return logger.debug('Denying fleet secretary from messaging themself');
 		}
+		if (isApproved) {
+			dmx.fireEvent(dmx.CHANNELS.DataHubVoteApproved);
```
Ilmeisesti tämä laukeaa silloin kun GM:t hyväksyvät äänestyksen ja muuttavat `isApproved == true`?  Onko mahdollista että myöhemmin jotain muuta muutettaisiin ja tämä triggeröityisi uudestaan?  Pitäisikö koittaa konditionalisoida siihen kun `isApproved` muuttuu false --> true?

nicou replied:
Oikein hyvä pointti 👍

plaa commented on src/rules/ship/jump.js:
```diff
@@ -229,10 +229,12 @@ function handleTransition(jump, currentStatus, previousStatus) {
 						shipLogger.warning(
 							`Jump crystal count is low (${jump_crystal_count} pcs)`
 						);
+						dmx.fireEvent(dmx.CHANNELS.LoraJumpCrystalsLow);
```
Tämä laukeaa jokaisen kerran kun kristalleja käytetään, ei vain silloin kun menee ensimmäisen kerran matalalle.  Onko tämä toivottu?  Voi olla tai ei  🤷

nicou replied:
Hyvä pointti, muutetaan kuulemma niin että tulee vain kun menee ensimmäisen kerran matalalle.

plaa commented:
En löydä missä on approve-nappi  🧐  Mutta voit mergetä kun olet katsonut kommentit onko tarpeen tehdä jotain.

nicou committed:
Move ship destruction DMX event

Makes more sense to do this in destroyShip() function instead of in the
router.

nicou committed:
Fix vote and post approval DMX triggers

Only fire the DMX event when status is changed to approved, not every
time an approved item is updated.

nicou committed:
Fix jump crystals low DMX event

Only fire the DMX event when jump crystal count initially dips under the
threshold instead of doing it every time the count is under the
threshold.
